src/memory/buffer_pool.h

```diff
@@ -4,74 +4,168 @@
+     * 
+     * @param bufferHandler BufferHandler for the allocated page.
+     * @return false if the alloc was successful, true otherwise.
+     */
+    ErrorCode alloc( BufferHandler& bufferHandler ) noexcept;
 
     /**
-     * Starts a transaction
-     * @return The id of the transaction
-     **/
-    transactionId_t beginTransaction();
+     * Releases a page from the Buffer Pool.
+     * 
+     * @param pId Page to release.
+     * @return false if the release was successful, true otherwise.
+     */
+    ErrorCode release( const pageId_t pId ) noexcept;
 
     /**
-     * Commits a transaction
-     * @param in tId The id of the transaction to commit
-     **/
-    void commitTransaction( const transactionId_t tId );
+     * Pins a page.
+     * 
+     * @param pId Page to pin.
+     * @param bufferHandler BufferHandler for the pinned page.
+     * @return false if the pin was successful, true otherwise.
+     */
+    ErrorCode pin( const pageId_t pId, BufferHandler& bufferHandler ) noexcept;
 
     /**
-     * Allocates a buffer
-     * @param The id of the transaction this allocation belongs to
-     * @return The id of the allocated buffer
-     **/
-    bufferId_t alloc( const transactionId_t tId ) noexcept;
+     * Unpins a page.
+     * 
+     * @param pId Page to unpin.
+     * @return false if the unpin was successful, true otherwise.
+     */
+    ErrorCode unpin( const pageId_t pId ) noexcept;
 
     /**
-     * Releases a buffer
-     * @param in bId The buffer to release
-     * @param in tId The id of the transaction 
-     **/
-    void release( const bufferId_t bId, const transactionId_t tId ) noexcept;
+     * Checkpoints the BufferPool to the storage.
+     * 
+     * @return false if the checkpoint was successful, true otherwise.
+     */
+    ErrorCode checkpoint() noexcept;
 
     /**
-     * Pins the buffer
-     * @param in bId The buffer to pin
-     * @param in tId The id of the transaction
-     * @return The handler of the buffer
-     **/
-    Buffer pin( const bufferId_t bId, const transactionId_t tId ) noexcept;
+     * Sets a page as dirty.
+     * 
+     * @param pId pageId_t of the page to be set as dirty.
+     */
+    void setPageDirty( pageId_t pId ) noexcept;
+
+  private:
 
     /**
-     * Checkpoints the BufferPool to the storage
-     **/
-    void checkpoint() noexcept;
+     * Get a pointer to the specified buffer slot.
+     * 
+     * @param bId bufferId_t of the buffer to get a pointer to.
+     * @return Pointer to bId buffer.
+     */
+    char* getBuffer( const bufferId_t bId ) noexcept;
+
+    /**
+     * Returns the bufferId_t of an empty buffer pool slot. In case none is free
+     * Clock Sweep algorithm is performed to evict a page.
+     * 
+     * @param bId bufferId_t of the free pool slot.
+     * @return false if all pages are pinned, true otherwise.
+     */
+    ErrorCode getEmptySlot( bufferId_t& bId ) noexcept;
+
+    struct bufferDescriptor {
```

Comment:
Can you define this struct at the top of the file?

Reply:
Solved.